src/primitives/decode/mod.rs

```diff
@@ -0,0 +1,384 @@
+//! - The separator, which is always "1".
+//! - The data part, which is at least 6 characters long and only consists of alphanumeric
+//!   characters excluding "1", "b", "i", and "o".
+//!
+//! ## Modules
+//!
+//! > We first describe the general checksummed base32 format called Bech32 and then define
+//! > Segregated Witness addresses using it.
+//!
+//! - mod.rs: Handles parsing the general checksummed base32 format.
+//! - segwit.rs: Handles parsing Segregated Witness addresses.
+//!
+//! [BIP-173]: <https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki>
+//! [BIP-350]: <https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki>
+
+pub mod segwit;
+
+use core::{fmt, iter, slice, str};
+
+use crate::primitives::checksum::{self, Checksum};
+use crate::primitives::gf32::Fe32;
+use crate::primitives::hrp::{self, Hrp};
+use crate::primitives::iter::{Fe32IterExt, FesToBytes};
+use crate::write_err;
+
+/// Separator between the hrp and payload (as defined by BIP-173).
+const SEP: char = '1';
+
+/// An HRP string that has been parsed and had the checksum validated.
+///
+/// Pre-parsing an HRP string only checks validity of the characters, it does not validate the
+/// checksum in any way - to validate convert to [`Parsed`].
+#[derive(Debug)]
+pub struct Parsed<'s> {
+    /// The human-readable part, guaranteed to be lowercase ASCII characters.
+    pub(crate) hrp: Hrp,
+    /// This is ASCII byte values of the parsed string, guaranteed to be valid bech32 characters,
+    /// with the checksum removed.
+    pub(crate) data: &'s [u8],
+}
+
+impl<'s> Parsed<'s> {
+    /// Parses an HRP string, without treating the first data character specially.
+    pub fn new<Ck: Checksum>(s: &'s str) -> Result<Self, Error> {
+        let unvalidated = Parsed::parse_unvalidated(s)?;
+        let ret = unvalidated.validate_checksum::<Ck>()?;
+        Ok(ret)
+    }
+
+    /// Returns the human-readable part.
+    pub fn hrp(&self) -> Hrp { self.hrp }
+
+    /// Returns an iterator over the byte data encoded by the HRP string (excluding the HRP and
+    /// checksum).
+    pub fn byte_iter(&self) -> ByteIter {
+        ByteIter { iter: AsciiToFe32Iter { iter: self.data.iter().copied() }.fes_to_bytes() }
+    }
+
+    /// Returns an iterate over field elements of the data encoded by the HRP string (excluding the
+    /// and checksum).
+    pub fn fe32_iter(&self) -> Fe32Iter {
+        Fe32Iter { iter: AsciiToFe32Iter { iter: self.data.iter().copied() } }
+    }
+
+    /// Parses an bech32 encode string and constructs a [`Parsed`] object.
+    ///
+    /// Checks for valid ASCII values, does not validate the checksum.
+    ///
+    /// The object returned by this function must must have `validate_checksum` called on it to
+    /// validate and checksum and maintain data invariant (remove checksum from parsed data).
+    fn parse_unvalidated(s: &'s str) -> Result<Self, Error> {
```

Comment:
In 5c9b549a3738cec82f633c14e274854b82edce45:

I'd call this `new_unvalidated` to match `new`.

Reply:
Oh, I see, it's not `pub`. No need to make change the doccomment then. Or the name, arguably.

Reply:
I've re-worked the constructors to get the test vectors all running - includes re-naming and making functions public.